### DISCUSSION TITLE
Print std type

### DISCUSSION THREAD

**Discussion by asolwa** [google/googletest#4121] [Community Help] [answered]
Hi!

Is there any way to provide custom printing function for **std types** e.g. std::future_status on version 1.11?
I tried methods suggested below:
[https://stackoverflow.com/questions/74007548/how-to-implement-a-custom-value-printer-for-googletest-for-std-enums](https://stackoverflow.com/questions/74007548/how-to-implement-a-custom-value-printer-for-googletest-for-std-enums)
[https://stackoverflow.com/questions/62054785/google-test-printto-for-stdsetstdstring](https://stackoverflow.com/questions/62054785/google-test-printto-for-stdsetstdstring)
but none of those works on version 1.11.

**Comment by Mahiuha** [comment, marked as answer]
In version 1.11 of Google Test, custom printing functions for standard types such as **std::future_status** can be implemented by creating a specialization of the **::testing::internal::UniversalPrinter** template class for the specific type in question.

Here is an example of how you might define a custom printing function for **std::future_status:**
```
namespace testing {
namespace internal {
template <>
class UniversalPrinter<std::future_status> {
 public:
  static void Print(const std::future_status& value, ::std::ostream* os) {
    if (value == std::future_status::ready) {
      *os << "ready";
    } else if (value == std::future_status::timeout) {
      *os << "timeout";
    } else if (value == std::future_status::deferred) {
      *os << "deferred";
    }
  }
};
}  // namespace internal
}  // namespace testing

```

You should put this code in a .cc file in your test project, and include it in your test targets.

This will allow you to use the **EXPECT_EQ** and **ASSERT_EQ** macros with **std::future_status** values and have the custom printing function used when the test fails and the values are printed.

It's worth noting that in later versions of GoogleTest, this functionality is provided by **TEST_P** macro.

**Reply by asolwa** [reply]
Unfortunatly it doen't works for std::chrono::system_clock::time_point, default UnivertaslPrinter is then selected.
How to print std::chrono::system_clock::time_point in this framework then?

**Reply by Mario-Klebsch** [reply]
std::chrono::system_clock::time_point probably is an alias for std::chrono::timepoint<>-specialization. I would try to implement UniversalPrinter<> for this type.

This works for me:
```
namespace testing {
namespace internal {

template<class Rep, class Period>
class UniversalPrinter<std::chrono::duration<Rep, Period>>
{
public:
	static void Print(const std::chrono::duration<Rep, Period>& value, ::std::ostream* os) {
		*os << value.count() << " * (" << Period::num << "/" << Period::den << ")s";
	}
};

template<class Clock, class Duration>
class UniversalPrinter<std::chrono::time_point<Clock, Duration>>
{
public:
	static void Print(const std::chrono::time_point<Clock, Duration>& value, ::std::ostream* os) {
		UniversalPrinter<Duration>::Print(value.time_since_epoch(), os);
		*os << " on clock " << ::testing::internal::GetTypeName<Clock>();
	}
};


}  // namespace internal
}  // namespace testing
```

```
auto now = std::chrono::system_clock::now();
auto x = std::chrono::system_clock::time_point{ std::chrono::system_clock::duration{ 100 } };
EXPECT_EQ(now, x);
```

```
error: Expected equality of these values:
  now
    Which is: 17872503648176711 * (1/10000000)s on clock std::chrono::system_clock
  x
    Which is: 100 * (1/10000000)s on clock std::chrono::system_clock

```
